tests/test_vectorizer.py

```diff
@@ -0,0 +1,585 @@
+        with con.cursor() as cur:
+            for query, expected in tests:
+                cur.execute(query)
+                actual = cur.fetchone()[0]
+                assert actual.keys() == expected.keys()
+                for k, v in actual.items():
+                    assert k in expected and v == expected[k]
+
+
+def test_formatting_config_python_string_template():
+    tests = [
+        (
+            """
+            select ai.formatting_config_python_string_template
+            ( array['size', 'shape']
+            , 'size: $size shape: $shape $chunk'
+            )
+            """,
+            {
+                "implementation": "python_string_template",
+                "columns": ["size", "shape"],
+                "template": "size: $size shape: $shape $chunk",
+            },
+        ),
+        (
+            """
+            select ai.formatting_config_python_string_template
+            ( array['color', 'weight', 'category']
+            , 'color: $color weight: $weight category: $category $chunk'
+            )
+            """,
+            {
+                "implementation": "python_string_template",
+                "columns": ["color", "weight", "category"],
+                "template": "color: $color weight: $weight category: $category $chunk",
+            },
+        ),
+    ]
+    with psycopg.connect(db_url("test")) as con:
+        with con.cursor() as cur:
+            for query, expected in tests:
+                cur.execute(query)
+                actual = cur.fetchone()[0]
+                assert actual.keys() == expected.keys()
+                for k, v in actual.items():
+                    assert k in expected and v == expected[k]
+
+
+def test_validate_formatting_config_python_string_template():
+    ok = [
+        """
+        select ai._validate_formatting_config_python_string_template
+        ( ai.formatting_config_python_string_template
+          ( array['color', 'weight']
+          , 'color: $color weight: $weight $chunk'
+          )
+        , 'public', 'thing'
+        )
+        """,
+    ]
+    bad = [
+        """
+        select ai._validate_formatting_config_python_string_template
+        ( ai.formatting_config_python_string_template
+          ( array['color', 'weight', 'height']
+          , 'color: $color weight: $weight height: $height $chunk'
+          )
+        , 'public', 'thing'
+        )
+        """,
+    ]
+    with psycopg.connect(db_url("test")) as con:
+        with con.cursor() as cur:
+            cur.execute("drop table if exists public.thing;")
+            cur.execute("create table public.thing (id int, color text, weight float)")
+            for query in ok:
+                cur.execute(query)
+                assert True
+            for query in bad:
+                with pytest.raises(psycopg.errors.RaiseException):
+                    cur.execute(query)
+            con.rollback()
+
+
+VECTORIZER_ROW = """
+{
+    "id": 1,
+    "config": {
```

Comment:
What about removing "config" and moving everything to the root?

Reply:
maybe. I had thought about that. I want to be able to add more config "things" later without changing the structure of the table, though. 🤔

Reply:
ok, I wasn't seeing it as a table, just as a json